keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,366 @@
+
+Use case 2:
+For some storage systems, volumes are always managed in a group. For these storage systems, they will have to create a group for a single volume if they need to implement a create volume function in Kubernetes. Providing a VolumeGroup API will be very convenient for them.
+
+Use case 3:
+Instead of taking individual snapshots one after another, VolumeGroup can be used as a source for taking a snapshot of all the volumes in the same volume group. This may be a storage level consistent group snapshot if the storage system supports it.  In any case, when used together with ExecutionHook, this group snapshot can be application consistent. For this use case, we will introduce another CRD GroupSnapshot.
+
+Use case 4:
+VolumeGroup can be used to manage group replication or consistency group replication if the storage system supports it. Note replication is out of scope for this proposal.  It is mentioned here as a potential future use case.
+
+Use case 5:
+VolumeGroup can be used to manage volume placement to either spread the volumes across storage pools or stack the volumes on the same storage pool. Related KEPs proposing the concept of storage pool for volume placement is as follows:
+  https://github.com/kubernetes/enhancements/pull/1353
+  https://github.com/kubernetes/enhancements/pull/1347
+We may not really need a VolumeGroup for this use case. A StoragePool is probably enough.  This is to be determined.
+
+Use case 6:
+VolumeGroup can also be used together with application snapshot. It can be a resource managed by the ApplicationSnapshot CRD.
+
+Use case 7: 
+Some applications may not want to use ApplicationSnapshot CRD because they don’t use Kubernetes workload APIs such as StatefulSet, Deployment, etc.  Instead, they have developed their own operators.  In this case it is more convenient to use VolumeGroup to manage persistent volumes used in those applications.
+
+### Goals
+
+* Provide an API to manage multiple volumes together in a group.
+* Provide an API to take a snapshot of a group of volumes.
+* Provide a design to facilitate volume placement using the group API (To be determined).
+* The group API should be generic and extensible so that it may be used to support other features in the future.
+
+### Non-Goals
+
+* A VolumeGroup may potentially be used to support replication group in the future, but providing design on replication group is not in the scope of this KEP. This can be discussed in the future.
+
+## Proposal
+
+This proposal introduces new CRDs VolumeGroup, VolumeGroupClass, and GroupSnapshot.
+
+Create new group:
+There are two ways to create a new VolumeGroup.
+* Create a group with existing volumes, either with a list of PVCs or using a selector.
+* Create an empty group first, then create each individual volume with group_id which will add a volume to the already created group.
+
```

Comment:
>Creating a group with an arbitrary set of existing pvcs may not be possible- they might be using disparate hardware resources (e.g. protection domains) that cannot do consistent snaps, or possibly they're on completely different storage systems or storage system types. I don't want k8s api to enforce something arbitrary like same StorageClass or something

@rbo54 if its not in PVC or namespace scoped, the proposal or API can have many challenges wrt to the security..etc.  For example, if we operate on PV, SC level the co-orindation/validation of allowed operations becomes a challengee IMO.

Reply:
I was reacting more to the list of use cases above, not proposing a new one. I'm just saying the driver may need insight into what the intended use for the volume group is in order to make placement decisions.

Reply:
I'm not proposing a new use case. But use case 1 (statefulset grouping) above the driver might be able to treat in a less restrictive way than use case 3 (snapshot consistency). With the current proposal though I don't see how the driver can distinguish between the use cases so it would always have to take the most restrictive view.